src/library/autodj/autodjprocessor.cpp

```diff
@@ -504,119 +578,130 @@ void AutoDJProcessor::playerPositionChanged(DeckAttributes* pAttributes,
                 // playing.
                 loadNextTrackFromQueue(rightDeck);
 
-                // Set crossfade thresholds for left deck.
-                calculateTransition(&leftDeck, &rightDeck);
+                // Note: calculateTransition() is called in playerTrackLoaded()
             } else {
-                // At least right Deck is playing
+                // At least right deck is playing
                 // Set crossfade thresholds for right deck.
-                calculateTransition(&rightDeck, &leftDeck);
+                calculateTransition(&rightDeck, &leftDeck, false);
             }
             emitAutoDJStateChanged(m_eState);
         }
         return;
     }
 
-    // In P1FADING/P2FADING states, we are waiting for the deck we are fading
-    // from (P1 or P2) to stop playing. Once we are playing and the other deck
-    // is not playing -- we switch the crossfader fully to this deck's side,
+    // In FADING states, we expect that both tracks are playing.
+    // Normally the the fading fromDeck stops after the transition is over and
+    // we need to replace it with a new track from the queue. In the rare case the
+    // toDeck stops first, we replace this one and stop the transition.
+    // Then we switch the crossfader fully to the new track side,
     // switch to IDLE mode and load the next track into the other deck.
-    if ((m_eState == ADJ_P1FADING && thisDeck.isRight()) ||
-        (m_eState == ADJ_P2FADING && thisDeck.isLeft())) {
+    if (m_eState == ADJ_LEFT_FADING || m_eState == ADJ_RIGHT_FADING) {
         // Once P1 or P2 has stopped switch out of fading mode to idle.
-        if (thisDeckPlaying && !otherDeckPlaying) {
-            // Force crossfader all the way to this side.
-            if (thisDeck.isLeft()) {
-                setCrossfader(-1.0, false);
+        if (!otherDeckPlaying) {
```

Comment:
This code does not always work well if you stop the toDeck(). I guess this function is called form the last Position change even though the deck is already stopped. In this case the code is executed as if where the fromDeck. 

What is a reasonable responds when one stops the too deck? In the 2.2 implementation, the AutoDJ continues fades to the stopped deck and stops. This is IMHO reasonable, because user takes control and we cannot predict why. Maybe he realizes that it was seeked wrong.   

Loading a new track and instantly start it with fader fully on the side is a bad option, because it will be notable.

Reply:
This was suddenly loading the next track when the to-deck was paused during a transition. I fixed this with cdb8148b13ad39d1342832b1ac796dead9cbabc3 I agree that continuing to fade to the stopped deck is reasonable. With the new change, AutoDJ will continue as if nothing happened whenever the user plays the to-deck again.